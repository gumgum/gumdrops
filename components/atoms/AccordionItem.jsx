import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import generateUID from '../utils/generateUID';
import charCodes from '../../constants/charCodes';

class AccordionItem extends Component {
    state = {
        isOpen: this.props.isOpen
    };

    uid = generateUID(this);

    toggleOpen = event => {
        const { type, charCode } = event;
        event.stopPropagation(); // don't want nested accordions to propagate events

        if (this.props.isLocked) return;

        if (
            type === 'keypress' &&
            (charCode === charCodes.SPACEBAR || charCode === charCodes.ENTER)
        ) {
            event.preventDefault(); // prevents page scroll from space key
            this.setState(({ isOpen }) => ({ isOpen: !isOpen }));
        } else if (type === 'click') {
            this.setState(({ isOpen }) => ({ isOpen: !isOpen }));
        }
    };

    render() {
        const { size, context, className, children, label, isLocked } = this.props;

        const { isOpen } = this.state;

        const rootClass = cx('gds-accordion__item', className, {
            [`gds-accordion__item--${context}`]: context,
            'gds-accordion__item--active': isOpen
        });

        const titleClass = cx('gds-accordion__item-title', {
            [`gds-accordion__item-title--${size}`]: size
        });

        const iconClass = cx('gds-accordion__item-icon', {
            [`gds-accordion__item-icon--${size}`]: size
        });

        const childClass = cx('gds-accordion__child-items', {
            [`gds-accordion__child-items--${size}`]: size
        });

        const newChildren = React.Children.map(children, child => {
            return React.cloneElement(child, {
                context,
                size
            });
        });

        const regionId = `AccordionItem-region-${this.uid}`;
        const labelId = `AccordionItem-label-${this.uid}`;

        return (
            <li
                aria-pressed={isOpen}
                aria-controls={regionId}
                aria-expanded={isOpen}
                className={rootClass}
                onClick={this.toggleOpen}
                onKeyPress={this.toggleOpen}
                role="button"
                tabIndex={isLocked ? -1 : 0}>
                <h4 id={labelId} className={titleClass}>
                    {label}
                </h4>
                {!isLocked && <span className={iconClass} />}
                <ul
                    aria-labelledby={labelId}
                    aria-hidden={!isOpen}
                    className={childClass}
                    id={regionId}
                    role="region">
                    {newChildren}
                </ul>
            </li>
        );
    }
}

AccordionItem.displayName = 'AccordionItem';

AccordionItem.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    context: PropTypes.string,
    isOpen: PropTypes.bool,
    isLocked: PropTypes.bool,
    label: PropTypes.string,
    size: PropTypes.string
};

export default AccordionItem;
