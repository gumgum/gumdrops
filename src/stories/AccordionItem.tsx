import React, { MouseEvent, KeyboardEvent, useState } from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';
import { keyCodes } from 'types';

type ToggleEvent = MouseEvent<HTMLLIElement> | KeyboardEvent<HTMLLIElement>;

export const AccordionItem: React.FC<AccordionItemProps> = ({
    isOpen: initialIsOpen,
    isLocked,
    label,
    size,
    color,
    className,
    children
}) => {
    const [isOpen, setIsOpen] = useState(initialIsOpen);

    const toggleOpen = (event: ToggleEvent): void => {
        event.stopPropagation(); // don't want nested accordions to propagate events
        event.preventDefault(); // prevents page scroll from space key
        const { type } = event;
        const { key } = event as KeyboardEvent;

        if (isLocked) return;
        const allowToggle =
            (type === 'keypress' && [keyCodes.space, keyCodes.enter].includes(key as keyCodes)) ||
            type === 'click';

        if (!allowToggle) return;

        setIsOpen(isOpen => !isOpen);
    };

    const rootClass = cx('gds-accordion__item', className, {
        [`gds-accordion__item--${color}`]: color,
        'gds-accordion__item--active': isOpen
    });

    const titleClass = cx('gds-accordion__item-title', {
        [`gds-accordion__item-title--${size}`]: size
    });

    const iconClass = cx('gds-accordion__item-icon', {
        [`gds-accordion__item-icon--${size}`]: size
    });

    const childClass = cx('gds-accordion__child-items', {
        [`gds-accordion__child-items--${size}`]: size
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size
        });
    });

    const regionId = `AccordionItem-region-${label}`;
    const labelId = `AccordionItem-label-${label}`;

    return (
        <li
            aria-pressed={isOpen}
            aria-controls={regionId}
            aria-expanded={isOpen}
            className={rootClass}
            onClick={toggleOpen}
            onKeyPress={toggleOpen}
            role="button"
            tabIndex={isLocked ? -1 : 0}>
            <h4 id={labelId} className={titleClass}>
                {label}
            </h4>
            {!isLocked && <span className={iconClass} />}
            <ul
                aria-labelledby={labelId}
                aria-hidden={!isOpen}
                className={childClass}
                id={regionId}
                role="region">
                {newChildren}
            </ul>
        </li>
    );
};
