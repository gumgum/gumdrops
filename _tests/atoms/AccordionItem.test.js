import renderer from 'react-test-renderer';
import React from 'react';
import AccordionItem from '../../components/atoms/AccordionItem';

test('Expect <AccordionItem> to render properly', () => {
    const props = {
        className: 'foo',
        label: 'Foo!',
        size: 'xs',
        context: 'danger'
    };
    const tree = renderer
        .create(
            <AccordionItem {...props}>
                <div>foo</div>
            </AccordionItem>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
