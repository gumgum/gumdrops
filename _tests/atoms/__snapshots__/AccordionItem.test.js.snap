// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <AccordionItem> to render 1`] = `
<AccordionItem
  className="foo"
  context="danger"
  label="Foo!"
  size="xs"
>
  <li
    aria-controls="AccordionItem-region-QWNjb3JkaW9uSXRlbQ==0"
    className="gds-accordion__item foo gds-accordion__item--danger"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <h4
      className="gds-accordion__item-title gds-accordion__item-title--xs"
      id="AccordionItem-label-QWNjb3JkaW9uSXRlbQ==0"
    >
      Foo!
    </h4>
    <span
      className="gds-accordion__item-icon gds-accordion__item-icon--xs"
    />
    <ul
      aria-hidden={true}
      aria-labelledby="AccordionItem-label-QWNjb3JkaW9uSXRlbQ==0"
      className="gds-accordion__child-items gds-accordion__child-items--xs"
      id="AccordionItem-region-QWNjb3JkaW9uSXRlbQ==0"
      role="region"
    >
      <div
        context="danger"
        key=".0"
        size="xs"
      >
        foo
      </div>
    </ul>
  </li>
</AccordionItem>
`;
